File: python/2276.py
# class CountIntervals:
#     __slots__ = 'left', 'right', 'l', 'r', 'cnt'

#     def __init__(self, l=1, r=10 ** 9):
#         self.left = self.right = None
#         self.l, self.r = l, r
#         self.cnt = 0

#     def add(self, l: int, r: int) -> None:
#         if self.r - self.l + 1 == self.cnt:
#             return
#         if l <= self.l and self.r <= r:
#             self.cnt = self.r - self.r + 1
#             return
#         mid = (self.l + self.r) // 2
#         if self.left == None: self.left = CountIntervals(self.l, mid)
#         if self.right == None: self.right = CountIntervals(mid+1, self.r)
#         if l <= mid: self.left.add(l, r)
#         if r >= mid+1: self.right.add(l, r)
#         self.cnt = self.left.cnt + self.right.cnt

#     def count(self) -> int:
#         return self.cnt



# Your CountIntervals object will be instantiated and called as such:
# obj = CountIntervals()
# obj.add(left,right)
# param_2 = obj.count()

from sortedcontainers import SortedDict


class CountIntervals:

    def __init__(self):
        self.d = SortedDict()
        self.cnt = 0

    def add(self, left: int, right: int) -> None:
        i = self.d.bisect_left(left)
        while i < len(self.d) and self.d.values()[i] <= right:
            r, l = self.d.items()[i]
            left = min(left, l)
            right = max(right, r)
            self.cnt -= r - l + 1
            self.d.popitem(i)
        self.cnt += right - left + 1
        self.d[right] = left

    def count(self) -> int:
        return self.cnt


# Your CountIntervals object will be instantiated and called as such:
# obj = CountIntervals()
# obj.add(left,right)
# param_2 = obj.count()

File: go/2276.go
package main

import "github.com/emirpasic/gods/trees/redblacktree"

// type CountIntervals struct {
// 	l, r, cnt   int
// 	left, right *CountIntervals
// }

// func Constructor() CountIntervals {
// 	return CountIntervals{l: 1, r: 1_000_000_000}
// }

// func (c *CountIntervals) Add(l int, r int) {
// 	if c.cnt == c.r-c.r+1 {
// 		return
// 	}
// 	if l <= c.l && c.r <= r {
// 		c.cnt = c.r - c.l + 1
// 		return
// 	}
// 	mid := (c.l + c.r) / 2
// 	if c.left == nil {
// 		c.left = &CountIntervals{l: c.l, r: mid}
// 	}
// 	if c.right == nil {
// 		c.right = &CountIntervals{l: mid + 1, r: c.r}
// 	}
// 	if l <= mid {
// 		c.left.Add(l, r)
// 	}
// 	if r >= mid+1 {
// 		c.right.Add(l, r)
// 	}
// 	c.cnt = c.left.cnt + c.right.cnt
// }

// func (c *CountIntervals) Count() int {
// 	return c.cnt
// }

/**
 * Your CountIntervals object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Add(left,right);
 * param_2 := obj.Count();
 */

type CountIntervals struct {
	*redblacktree.Tree
	cnt int
}

func Constructor() CountIntervals {
	return CountIntervals{redblacktree.NewWithIntComparator(), 0}
}

func (c *CountIntervals) Add(left int, right int) {
	for node, _ := c.Ceiling(left); node != nil && node.Value.(int) <= right; node, _ = c.Ceiling(left) {
		l, r := node.Value.(int), node.Key.(int)
		if l < left {
			left = l
		}
		if r > right {
			right = r
		}
		c.cnt -= r - l + 1
		c.Remove(r)
	}
	c.cnt += right - left + 1
	c.Put(right, left)
}

func (c *CountIntervals) Count() int {
	return c.cnt
}
